libfwupdplugin/fu-pefile-firmware.c

```diff
@@ -118,33 +140,89 @@ fu_pefile_firmware_parse(FuFirmware *firmware,
 	for (guint idx = 0; idx < nr_sections; idx++) {
 		guint32 sect_length = 0x0;
 		guint32 sect_offset = 0x0;
-		guint8 name[8] = {0x0};
+		guint32 zeroes = 0x0;
 		g_autofree gchar *sect_id = NULL;
 		g_autoptr(FuFirmware) fw_sect = NULL;
 		g_autoptr(GBytes) sect_blob = NULL;
 
-		/* read the section name */
-		if (!fu_memcpy_safe(name,
-				    sizeof(name),
-				    0x0,
-				    buf,
-				    bufsz,
-				    hdr_offset,
-				    sizeof(name),
-				    error))
-			return FALSE;
-		sect_id = fu_strsafe((const gchar *)name, sizeof(name));
-		if (sect_id == NULL) {
-			g_set_error_literal(error,
-					    FWUPD_ERROR,
-					    FWUPD_ERROR_INVALID_FILE,
-					    "no section name");
+		if (!fu_memread_uint32_safe(buf,
+					    bufsz,
+					    hdr_offset,
+					    &zeroes,
+					    G_LITTLE_ENDIAN,
+					    error))
 			return FALSE;
+
+		if (zeroes == 0) {
+			guint32 str_idx = 0x0;
+
+			if (!fu_memread_uint32_safe(buf,
+						    bufsz,
+						    hdr_offset + 4,
+						    &str_idx,
+						    G_LITTLE_ENDIAN,
+						    error))
+				return FALSE;
+
+			sect_id = fu_strsafe((const gchar *)buf + string_table + str_idx,
+					     bufsz - string_table);
+			if (sect_id == NULL) {
+				g_set_error_literal(error,
+						    FWUPD_ERROR,
+						    FWUPD_ERROR_INVALID_FILE,
+						    "no section name");
+				return FALSE;
+			}
+		} else {
+			guint8 name[8] = {0x0};
+
+			/* read the section name */
+			if (!fu_memcpy_safe(name,
+					    sizeof(name),
+					    0x0,
+					    buf,
+					    bufsz,
+					    hdr_offset,
+					    sizeof(name),
+					    error))
+				return FALSE;
+
+			sect_id = fu_strsafe((const gchar *)name, sizeof(name));
+
+			if (sect_id == NULL) {
+				g_set_error_literal(error,
+						    FWUPD_ERROR,
+						    FWUPD_ERROR_INVALID_FILE,
+						    "no section name");
+				return FALSE;
+			}
+
+			if (sect_id[0] == '/') {
+				guint64 str_idx = 0x0;
+
+				if (!fu_strtoull(sect_id + 1, &str_idx, 0, G_MAXUINT32, error))
+					return FALSE;
+
+				g_steal_pointer(&sect_id);
+
+				sect_id = fu_strsafe((const gchar *)buf + string_table + str_idx,
+						     bufsz - string_table);
+
+				if (sect_id == NULL) {
+					g_set_error_literal(error,
+							    FWUPD_ERROR,
+							    FWUPD_ERROR_INVALID_FILE,
+							    "no section name");
+					return FALSE;
+				}
+			}
 		}
 
 		/* create new firmware */
 		if (g_strcmp0(sect_id, ".sbom") == 0) {
 			fw_sect = fu_coswid_firmware_new();
+		} else if (g_strcmp0(sect_id, ".sbatlevel") == 0) {
+			fw_sect = fu_sbatlevel_section_firmware_new();
```

Comment:
I don't have any files with a .sbatlevel section -- i only have .sbat -- why?

Reply:
On the shim binary? Probably because you have an old one. `.sbatlevel` is only on shim (as well as `.sbat`). Other binaries have `.sbat` only. Older versions of shim should not have `.sbatlevel`. This was introduced last year.

Reply:
It is only for introspection. Originally to calculate PCR values I think. But the "previous" entry gives us what shim will install as sbatlevel. Which is the value we want to compare to. `.sbat` will usually be newer that the current sbatlevel.

Reply:
Also note that this PR allows installation if shim has no `.sbatlevel`. Because it will not brick the system.

Reply:
Hmm, on Fedora shim I don't see it at all:

	[hughsie@hughsie-work build (wip/hughsie/pefile %)]$ ninja src/fwupdtool && sudo ./src/fwupdtool firmware-parse /boot/efi/EFI/fedora/shimx64.efi pefile 
	ninja: no work to do.
	[sudo] password for hughsie: 
	Loading…                 [************************************** ]
	<firmware gtype="FuPefileFirmware">
	  <data size="0xe7218">MZ.</data>
	  <firmware>
	    <id>.eh_frame</id>
	    <offset>0x400</offset>
	    <size>0x1db1c</size>
	    <data size="0x1db1c">
	    </data>
	  </firmware>
	  <firmware>
	    <id>.text</id>
	    <idx>0x1</idx>
	    <offset>0x1e000</offset>
	    <size>0x5e663</size>
	    <data size="0x5e663">H...QRH.=....H.5</data>
	  </firmware>
	  <firmware>
	    <id>.reloc</id>
	    <idx>0x2</idx>
	    <offset>0x7c800</offset>
	    <size>0xa</size>
	    <data size="0xa">
	    </data>
	  </firmware>
	  <firmware>
	    <id>.data.ident</id>
	    <idx>0x3</idx>
	    <offset>0x7ca00</offset>
	    <size>0x49</size>
	    <data size="0x49">UEFI SHIM.$Versi</data>
	  </firmware>
	  <firmware>
	    <id>.data</id>
	    <idx>0x4</idx>
	    <offset>0x7cc00</offset>
	    <size>0x2d5b4</size>
	    <data size="0x2d5b4">S</data>
	  </firmware>
	  <firmware>
	    <id>.vendor_cert</id>
	    <idx>0x5</idx>
	    <offset>0xaa200</offset>
	    <size>0x37c</size>
	    <data size="0x37c">l.</data>
	  </firmware>
	  <firmware>
	    <id>.dynamic</id>
	    <idx>0x6</idx>
	    <offset>0xaa600</offset>
	    <size>0x100</size>
	    <data size="0x100">
	    </data>
	  </firmware>
	  <firmware>
	    <id>.rela</id>
	    <idx>0x7</idx>
	    <offset>0xaa800</offset>
	    <size>0x1b468</size>
	    <data size="0x1b468">
	    </data>
	  </firmware>
	  <firmware>
	    <id>.sbat</id>
	    <idx>0x8</idx>
	    <offset>0xc5e00</offset>
	    <size>0xdf</size>
	    <data size="0xdf">sbat,1,SBAT Vers</data>
	  </firmware>
	</firmware>

Reply:
I have checked. Fedora 38 has shim 15.6. Both .sbatlevel and the new revocations have been added to 15.7. shim 15.6 is always safe to install (might not boot though).

So what you see is expected. And this branch will not prevent from installing shim from Fedora 38.

Reply:
Aha, gotcha. Mind if I convert your parsing to a generic `FuCsvFirmware` object and then we layer on top of that too? I think it'll be useful in other places too.

Reply:
Sure. We could even use that for `.sbat` if one day we want to test a binary is revoked by current sbatlevel.

Reply:
Exactly my thoughts too.

Reply:
Should I convert the .sbatlevel parsing to use rustgen generated headers?

Reply:
sure, makes sense.

Reply:
(the part that extract the "previous" and "latest" entries, I mean, not csv parsing)